crates/ibc/src/core/ics03_connection/handler/conn_open_ack.rs

```diff
@@ -4,85 +4,129 @@ use crate::core::ics03_connection::connection::{ConnectionEnd, Counterparty, Sta
-            conn_end.delay_period(),
-        )
-    };
+        {
+            let counterparty_connection_id = self_connection_end
+                .counterparty()
+                .connection_id()
+                .ok_or_else(Error::invalid_counterparty)?;
+            let counterparty_expected_connection_end = ConnectionEnd::new(
+                State::TryOpen,
+                self_connection_end.counterparty().client_id().clone(),
+                Counterparty::new(
+                    self_connection_end.client_id().clone(), // The local client identifier.
+                    Some(msg.connection_id.clone()), // This chain's connection id as known on counterparty.
+                    ctx.commitment_prefix(),         // Local commitment prefix.
+                ),
+                vec![msg.version.clone()],
+                self_connection_end.delay_period(),
+            );
 
-    // 2. Pass the details to the verification function.
-    verify_proofs(
-        ctx,
-        msg.client_state,
-        msg.proofs.height(),
-        &conn_end,
-        &expected_conn,
-        &msg.proofs,
-    )?;
+            client_state
+                .verify_connection_state(
+                    msg.proofs_height,
+                    self_connection_end.counterparty().prefix(),
+                    &msg.proof_connection_end,
+                    consensus_state.root(),
+                    counterparty_connection_id,
+                    &counterparty_expected_connection_end,
+                )
+                .map_err(Error::verify_connection_state)?;
+        }
 
-    output.log("success: connection verification passed");
+        client_state
+            .verify_client_full_state(
+                msg.proofs_height,
+                self_connection_end.counterparty().prefix(),
+                &msg.proof_client_state,
+                consensus_state.root(),
+                self_connection_end.counterparty().client_id(),
+                msg.client_state,
+            )
+            .map_err(|e| {
+                Error::client_state_verification_failure(self_connection_end.client_id().clone(), e)
+            })?;
+
+        let expected_consensus_state = ctx.host_consensus_state(msg.consensus_height)?;
+        client_state
+            .verify_client_consensus_state(
+                msg.proofs_height,
+                self_connection_end.counterparty().prefix(),
+                &msg.proof_consensus_state,
+                consensus_state.root(),
+                self_connection_end.counterparty().client_id(),
+                msg.consensus_height,
+                expected_consensus_state.as_ref(),
+            )
+            .map_err(|e| Error::consensus_state_verification_failure(msg.proofs_height, e))?;
+    }
 
     let result = ConnectionResult {
         connection_id: msg.connection_id,
         connection_id_state: ConnectionIdState::Reused,
-        connection_end: conn_end,
+        connection_end: ConnectionEnd::new(
```

Comment:
I find that the setters we used previously we're easier to follow because it was clear what we changed/updated in the connection-end, maybe something like the following or extracted as a function (say `update_connection_end()`) -> 
```rust
let new_connection_end = {
    let mut counterparty = self_connection_end.counterparty().clone();
    counterparty.connection_id = Some(msg.counterparty_connection_id.clone());

    let mut connection = self_connection_end;
    connection.set_state(State::Open);
    connection.set_version(msg.version.clone());
    connection.set_counterparty(counterparty);
    connection
};
```

Reply:
Agreed [53b6e90](https://github.com/cosmos/ibc-rs/pull/170/commits/53b6e90ec0ec3e4160be558db2a8fe87b6b464fa)